### ISSUE TITLE
Add catalog and extension views

### ISSUE THREAD

**Issue by jcfr** [KitwareMedical/slicer-extensions-webapp#2]
# Catalog view

This view will display:
* a list of extensions associated with a version of Slicer described using slicer revision (called application revision), operating system and architecture.
* a list of categories on the left side for filtering down the list by extension category (more details below)

## Proposed routes

```
<server_url>/#/catalog/:app_revision/:os
<server_url>/#/catalog/:app_revision/:os/:arch
```

Using `<server_url>/#/catalog/:app_revision/:os` should "redirect" to `<server_url>/#/catalog/:app_revision/:os/:arch`

Route parameters:
* `app_revision`:
  * required
  * integer
  * This uniquely identify a version of the source used to build the application (it is monotonically  increasing)
* `os`:
  * required
  * accepted values `['macos', 'linux', 'macos']`
* `arch`: 
  * optional: default value is `amd64`
  * accepted values `['i386', 'amd64']`

## Relevant API calls

List of extensions can directly be retrieved from the backend doing the following:

http://192.168.113.208:8080/api/v1/app/5aa3da37f3eb080001497b3d/extension?app_revision=27201&os=macosx&arch=amd64&sort=created&sortdir=-1

## Initial design

![image](https://user-images.githubusercontent.com/219043/85339212-a9d8f300-b4b1-11ea-9371-9062ce136f4e.png)

![image](https://user-images.githubusercontent.com/219043/85339866-0092fc80-b4b3-11ea-9035-4572f4130c76.png)

### How to generate the category list ? 

In this first implementation, let's consider it is a flat list of categories associated with the set extension currently displayed in the catalog

If could be generated using code similar to this: https://github.com/Pierre-Assemat/Vuejs-Project/blob/77605bb2c4bf69128d1e381cd9dfbe871e45ec32/src/views/extensions-catalog/ExtensionsCatalogView.vue#L59-L83

In the second iteration, we will generate a tree (this is captured in #3)

### How to integrate category in the router ? 

* If we wanted to expose the category in the router, may be we could do something like this:

```
<server_url>/#/catalog/:app_revision/:os/:arch/[:category]
```

# Extension view

## Proposed routes

```
<server_url>/#/view/:extension_name/:app_revision/:os
<server_url>/#/view/:extension_name/:app_revision/:os/:arch
```

## Relevant API calls

Metadata specific to a given extension can directly be retrieved from the backend doing the following:

http://192.168.113.208:8080/api/v1/app/5aa3da37f3eb080001497b3d/extension?extension_id=5afff7a479e39e0001040ff0&sort=created&sortdir=-1

where the `extension_id` is `_id` (which corresponds to an idem id) returned in the API call used to generated the catalog view.

## Initial design

![image](https://user-images.githubusercontent.com/219043/85340119-83b45280-b4b3-11ea-84e1-5ad9d48a1e55.png)

# Frontend configuration

* `app_id`
* `backend_url` currently http://192.168.113.208:8080 but ultimately it will be `https://slicer-packages.kitware.com` when deployed in production. When developing, it could also be a local URL

# Notes

A revision can be associated with either a _stable_ (also called _release_) version or a _preview_ (also called _draft_) version of the application.

Since we could potentially deploy other instance of the frontend associated with different `app_id` (e.g a Slicer-based commercial application), let's  keep `app_revision` in  the code instead of `slicer_revision`

**Comment by subdavis** [comment]
Thanks, JC.  Getting started on this today.  Your local girder instance is fine for getting started, but there should be docs for setting up https://github.com/girder/slicer_package_manager as a development server.

**Comment by subdavis** [comment]
https://github.com/Pierre-Assemat/Vuejs-Project/blob/77605bb2c4bf69128d1e381cd9dfbe871e45ec32/src/views/extensions-catalog/ExtensionsCatalogView.vue#L59-L83

This references `extension.meta.category` but that field doesn't exist on http://192.168.113.208:8080/api/v1/app/5aa3da37f3eb080001497b3d/extension?app_revision=27201&os=macosx&arch=amd64&sort=created&sortdir=-1

@jcfr am I missing something?

**Comment by jcfr** [comment]
> but there should be docs for setting up https://github.com/girder/slicer_package_manager as a development server.

Agreed. I will be working on updating https://slicer-package-manager.readthedocs.io/en/latest/developer_guide.html

> `extension.meta.category` 

Thanks for the report

I noticed when writing the issue and was initially thinking it would be there for newer packages but is it is not the case:
* last night upload batch associated with revision 29173 is also missing it - see [here](http://192.168.113.208:8080/api/v1/app/5aa3da37f3eb080001497b3d/extension?app_revision=29173&os=macosx&arch=amd64&sort=created&sortdir=-1)
* invocation of the python client used for uploading doesn't specify it either- see [here](https://github.com/Slicer/Slicer/blob/02c06341b4c66b0deda953fb01379c8e58aadefd/Extensions/CMake/SlicerExtensionPackageAndUploadTarget.cmake#L330-L341)
* documentation of the  client does not mention it - see [here](https://slicer-package-manager.readthedocs.io/en/latest/commands_shell.html#upload-a-new-extension)

Good news is that:
* the [slicer_package_manager_client.uploadExtension](https://github.com/girder/slicer_package_manager/blob/addd15cd74d6a5bbbe5fb7a5338d76136926817c/python_client/slicer_package_manager_client/__init__.py#L229-L258) function support it (along with metadata like `screenshots`, ...)
* the CLI also accepts the `--category` parameter. See [slicer_package_manager_client/cli.py](https://github.com/girder/slicer_package_manager/blob/addd15cd74d6a5bbbe5fb7a5338d76136926817c/python_client/slicer_package_manager_client/cli.py#L451-L453)

Todo:
* [ ] Fix documentation. See https://github.com/girder/slicer_package_manager/issues/75 
* [ ] Update invocation of python client. See https://github.com/Slicer/Slicer/issues/5010
* [ ] Add category to existing packages. See https://github.com/girder/slicer_package_manager/issues/76

**Comment by subdavis** [comment]
Trivial first pass.

Right now, there's no info coming out of the backend about 1) extension author 2) thumbnail image 3) description so I can't really add that.  Also need category info before I can do filtering.

I'll see about doing the slicer integration parts next.

![Screenshot from 2020-06-24 14-38-22](https://user-images.githubusercontent.com/4214172/85613953-6b197900-b628-11ea-83d9-35ec1cc0ea5a.png)

**Comment by jcfr** [comment]
Looks great ! 

Missing metadata have been added (description, icon_url, category, homepage, screenshots)

**Comment by subdavis** [comment]
## Dummy details page added

![Screenshot from 2020-06-26 15-54-58](https://user-images.githubusercontent.com/4214172/85896189-0bf16b00-b7c6-11ea-8545-5e1b2ccc05d7.png)

## Extension list updated

![Screenshot from 2020-06-26 15-55-47](https://user-images.githubusercontent.com/4214172/85896190-0bf16b00-b7c6-11ea-8e19-f61ca6b7e25c.png)

* No thought has been put into page transitions or category change transitions 
* Download button does nothing
* The route for details looks different than you asked:  `/view/:name` where `name` is an extension name.  I like how this makes the URL human-readable rather than including some kind of girder BSON id that doesn't tell a human anything.
* screenshots missing
* Author info missing.  IDK if I should try and look that up using the girder user ID or what.

Aside from those caveats, I think this issue is mostly done.

**Comment by jcfr** [comment]
> The route for details

With the updated route, how would the user select operating system, arch and app version ?

 > Author info missing. IDK if I should try and look that up using the girder user ID or what.

List of contributors is purely descriptive and not related to Girder users.

**Comment by subdavis** [comment]
> With the updated route, how would the user select operating system, arch and app version ?

This application appears to be organized such that every OS/arch/appversion has a different and unique extension.

http://192.168.113.208:8080/api/v1/app/5aa3da37f3eb080001497b3d/extension?extension_name=27201_quantitativereporting_macosx_amd64_2ef8580&sort=created&sortdir=-1

This follows from your example of how to get information on a specific example above.  I don't see a way in https://github.com/girder/slicer_package_manager/blob/master/server/api/app.py#L424-L441 to query for an extension by name where the os, arch, and version are able to be specified separately.

In other words, I don't see a way to say "Show me ALL QuantitativeReporting extensions of any os, arch, version".  I may be misreading the API.

> List of contributors is purely descriptive and not related to Girder users.

Alright, I'll leave this information out of the app.

**Comment by subdavis** [comment]
NVM, figured it out... http://192.168.113.208:8080/api/v1/app/5aa3da37f3eb080001497b3d/extension?baseName=QuantitativeReporting&sort=created&sortdir=-1

I was missing `baseName`

I'll fix routing.

**Comment by jcfr** [comment]
> I'll fix routing.

:pray: 


> Download button does nothing

See https://github.com/KitwareMedical/slicer-extensions-webapp/issues/6

**Comment by jcfr** [comment]
After locally building the application:

```
export VUE_APP_APP_ID=5aa3da37f3eb080001497b3d
export VUE_APP_BASE_URL=http://192.168.113.208:8080/api/v1
yarn
yarn serve
```

Browsing to the following pages returns a blank page:

http://localhost:8080/#/catalog/All/29189/macos
http://localhost:8080/#/catalog/all/29189/macos

**Comment by subdavis** [comment]
@jcfr  There is no `macos`: https://github.com/girder/slicer_package_manager/blob/master/server/api/app.py#L435

It's `macosx`.

**Comment by jcfr** [comment]
I tried these without success:

* http://localhost:8080/#/catalog/All/29189/macosx
* http://localhost:8080/#/catalog/all/29189/macosx

(Note that there are extensions for `29189` and `macosx`, see [here](http://192.168.113.208:8080/#collection/5a8c800cd7a3fb0001b37cdd/folder/5ef853e34db7a400019d638a))

Do you have an example of working URLs you could share ?

**Comment by subdavis** [comment]
@jcfr Looks like the server is fixed.  This app uses history mode, so the correct url is:

* http://localhost:8080/catalog/all/29189/macosx

Note that `all` is case-insensitive, so either will work.

**Comment by jcfr** [comment]
It works :tada:
